pkg/bdb/hash_page.go

```diff
@@ -0,0 +1,119 @@
+package bdb
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/go-restruct/restruct"
+	"io"
+	"os"
+)
+
+// source: https://github.com/berkeleydb/libdb/blob/5b7b02ae052442626af54c176335b67ecc613a30/src/dbinc/db_page.h#L259
+type HashPage struct {
+	LSN            [8]byte `struct:"[8]byte"` /* 00-07: LSN. */
+	PageNo         uint32  `struct:"uint32"`  /* 08-11: Current page number. */
+	PreviousPageNo uint32  `struct:"uint32"`  /* 12-15: Previous page number. */
+	NextPageNo     uint32  `struct:"uint32"`  /* 16-19: Next page number. */
+	NumEntries     uint16  `struct:"uint16"`  /* 20-21: Number of items on the page. */
+	FreeAreaOffset uint16  `struct:"uint16"`  /* 22-23: High free byte page offset. */
+	TreeLevel      uint8   `struct:"uint8"`   /*    24: Btree tree level. */
+	PageType       uint8   `struct:"uint8"`   /*    25: Page type. */
+}
+
+func ParseHashPage(data []byte) (*HashPage, error) {
+	var hashPage HashPage
+
+	err := restruct.Unpack(data, binary.LittleEndian, &hashPage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack: %w", err)
+	}
+
+	return &hashPage, nil
+}
+
+func HashPageValueContent(db *os.File, pageData []byte, hashPageIndex uint16, pageSize uint32) ([]byte, error) {
+	// the first byte is the page type, so we can peek at it first before parsing further...
+	valuePageType := pageData[hashPageIndex]
+
+	// only HOFFPAGE page types have data of interest
+	if valuePageType != HashOffIndexPageType {
+		return nil, fmt.Errorf("only HOFFPAGE types supported (%+v)", valuePageType)
+	}
+
+	hashOffPageEntryBuff := pageData[hashPageIndex : hashPageIndex+HashOffPageSize]
+
+	entry, err := ParseHashOffPageEntry(hashOffPageEntryBuff)
+	if err != nil {
+		return nil, err
+	}
+
+	var hashValue []byte
+
+	for currentPageNo := entry.PageNo; currentPageNo != 0; {
+		pageStart := pageSize * currentPageNo
+
+		_, err := db.Seek(int64(pageStart), io.SeekStart)
+		if err != nil {
+			return nil, fmt.Errorf("failed to seek to HashPageValueContent (page=%d): %w", currentPageNo, err)
+		}
+
+		currentPageBuff, err := slice(db, int(pageSize))
+		if err != nil {
+			return nil, fmt.Errorf("failed to read page=%d: %w", currentPageNo, err)
+		}
+
+		currentPage, err := ParseHashPage(currentPageBuff)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse page=%d: %w", currentPageNo, err)
+		}
```

Comment:
I meant

```suggestion
		currentPage, err := ParseHashPage(currentPageBuff)
		if err != nil {
			return nil, fmt.Errorf("failed to parse page=%d: %w", currentPageNo, err)
		}
		if currentPage.PageType != OverflowPageType {
			continue
		}
```

```
	OverflowPageType PageType = 7
```

https://github.com/berkeleydb/libdb/blob/5b7b02ae052442626af54c176335b67ecc613a30/src/dbinc/db_page.h#L43

Reply:
Ahh, I was interpreting as a go type and was confused, but agreed, asserting that the page type is indeed an overflow page is correct 👍